### ISSUE TITLE
[YSQL] Deadlock detection does not work at any of the available three isolation levels

### ISSUE THREAD

**Issue by bllewell** [yugabyte/yugabyte-db#15695]
Jira Link: [DB-5051](https://yugabyte.atlassian.net/browse/DB-5051)

### Environment

The tests were done both in a single-node YugabyteDB Version YB-2.17.0.0 cluster (hereinafter _YB_) and in a PostgreSQL Version 11 cluster (hereinafter _PG_). Both clusters were installed in a Ubuntu VM.

**IMPORTANT:** You might need to set a _gflag_ to allow the use of the _"read committed"_ isolation level, thus:

```
--yb_enable_read_committed_isolation=true
```

### Summary

The possibility of deadlock detection is discussed in no end of writings about RDBMSs. Here's a well-known text-book illustration of how deadlock can occur:

```
-- RED session                                  -- BLUE session
---------------------------------               ---------------------------------
start transaction
  isolation level read committed;
set deadlock_timeout = ...
update t set v = v+3 where k = 1;
                                                start transaction
                                                  isolation level read committed;
                                                set deadlock_timeout = ...
                                                update t set v = v+5 where k = 2;
update t set v = v+3 where k = 2;
                                                update t set v = v+5 where k = 1;
```

- The _BLUE_ session has implicitly locked the row with _"k = 2"_ (by updating it) when the _RED_ session issues an _update_ for the same row. So _RED_ must wait for _BLUE_ to commit (or rollback) — and _RED_ is therefore not yet listening for the next SQL statement.

- The _RED_ session has implicitly locked the row with _"k = 1"_ (by updating it) when the _BLUE_ session issues an _update_ for the same row. So _BLUE_ must wait for _RED_ to commit (or rollback) — and _BLUE_ is therefore not yet listening for the next SQL statement.

This mutual waiting is called _deadlock_. And every practical RDMBS must detect a deadlock _behind the scenes_, choose a victim, abort it, and issue a suitable error message. **PG does this properly and reliably**. The error text contains something like this:

```
Process 1569952 waits for ShareLock on transaction 15912; blocked by process 1569517.
Process 1569517 waits for ShareLock on transaction 15913; blocked by process 1569952.
```

And if you get the PID for each session before starting, with the _pg_backend_pid()_ built-in function, then you'll see that the values that it reports form each session match the error text.

**In YB, the deadlock is simply not detected—and the mutual wait continues indefinitely.**

Moreover, if you abort the waiting statement in one session with _Ctrl-C_, then the other session continues to wait. So you must abort that too.

You get this outcome with each of the three available isolation levels:
- read committed
- repeatable read
- serializable

### Testcase

>  **NOTE:** The testcase uses the superuser _yugabyte_. This is normally terrible practice. I developed it using an ordinary user with no special privileges, according to proper practice. But this clutters the scripts. And the errors that I saw reproduce when you use _yugabyte_. If you want to run the tests in _PG_ to see the proper, bug-free, behavior, then the simplest thing is to (once-off) create a superuser called _yugabyte_, and a database called _yugabyte_, there too before then using the provided scripts "as is".

The good outcome, in PG, is identical at all the levels. And the buggy outcome, in YB, is also identical at all the levels.

You'll see two _psql_ shortcuts in use below — _:cr_u0_ and _:u0_.

- The first invokes a script (not included in this report) that drops and recreates an ordinary user, say _u0_, in any suitable database, with a suitable schema first in the _search_path_.
- The second simply connects as the test user in the chosen database.

Each session, _RED_, and _BLUE_, will remain remain in the session that was started for it for the rest of the test.

Set up with this:

```
\c yugabyte yugabyte
set client_min_messages = warning;
drop database if exists d0;
create database d0;
\c d0 yugabyte
set client_min_messages = warning;

drop table if exists t cascade;
create table t(k int primary key, v int not null);
insert into t(k, v) values (1, 10), (2, 20), (3, 30), (4, 40), (5, 50);
select * from t order by 1;

drop procedure if exists set_deadlock_timeout(int) cascade;
create procedure set_deadlock_timeout(ms in int)
  security definer
  language plpgsql
as $body$
begin
  -- Must be done as a superuser.
  execute format('set deadlock_timeout = %L', ms::text);
end;
$body$;
```

Now step through the following scenario by hand using two sessions. (I used terminal windows with, respectively, _RED_ and _BLUE_ backgrounds improve the experiment's usability.)

```
-- RED                                              -- BLUE
--------------------------------------------------  --------------------------------------------------
\c d0 yugabyte
\set VERBOSITY verbose
select pg_backend_pid();
start transaction
  isolation level read committed;
call set_deadlock_timeout(3000); -- 3 seconds
update t set v = v+3 where k = 1;
                                                -- Start the BLUE session now.
                                                \c d0 yugabyte
                                                set client_min_messages = warning;
                                                \set VERBOSITY verbose
                                                select pg_backend_pid();
                                                start transaction
                                                  isolation level read committed;
                                                call set_deadlock_timeout(1000); -- 1 second
                                                update t set v = v+5 where k = 2;
update t set v = v+3 where k = 2;
-- RED is now waiting on BLUE
                                                    update t set v = v+5 where k = 1;
                                                    -- In PG, BLUE now gets an immediate error [1]
                                                    -- In YB, BLUE waits forever (and so does RED). [2]
                                                    -- So in YB, do Ctrl-C. And the txn is broken.
-- In PG, RED is now freed up.
-- In YB, RED is is still waiting.
-- So in YB, do Ctrl-C. And the txn is broken.
commit; -- (Means "rollback" in YB.)
-- In PG, RED sees what it asked for.
-- In YB, the rows are unchanged (of course).
select * from t order by 1;
                                                    commit; -- (Means "rollback". It's the only choice.)
                                                    -- BLUE sees the same as RED (of course).
                                                    select * from t order by 1;
```

Here is the error that _BLUE_ sees in _PG_.

```
ERROR:  deadlock detected
DETAIL:  Process 1569952 waits for ShareLock on transaction 15912; blocked by process 1569517.
Process 1569517 waits for ShareLock on transaction 15913; blocked by process 1569952.
HINT:  See server log for query details.
CONTEXT:  while updating tuple (0,1) in relation "t"
```

**NOTES:**

[1] _set deadlock_timeout_ is invoked immediately following _start transaction_ in both the _RED_ and the _BLUE_ sessions. Notice that _RED_ sets _three_ seconds and _BLUE_ sets just _one_ second. In PG, this ensures that, in _PG_, it's BLUE that suffers the _deadlock detected_ error. You can swap the values to ensure that it's _RED_ that suffers the error.

[2[ However, in _YB_, the _deadlock_timeout_ seems not to be consulted. The two sessions simply wait indefinitely on each other.

[DB-5051]: https://yugabyte.atlassian.net/browse/DB-5051?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by bllewell** [comment]
This might share the same root cause with [Issue #15675](https://github.com/yugabyte/yugabyte-db/issues/15675)

**"server closed the connection unexpectedly" errors in two-concurrent-session deferred FKs test **

**Comment by bllewell** [comment]
The root cause of this issue might explain the buggy behavior that [Issue #15717]( https://github.com/yugabyte/yugabyte-db/issues/15717] reports.

**Comment by robertsami** [comment]
to @shishir2001-yb to close
